File: Java/509. Fibonacci Number.java
class Solution {
    int [] dp;
    public int fib(int n) {
        dp = new int[n+1];
        return solve(n);
    }
    public int solve(int n){
        if (n==0) return 0;
        if (n==1) return 1;
        if (dp[n]!=0) return dp[n];
        else{
            dp[n] = solve(n-1) + solve(n-2);
            return dp[n];
        }
    }
}

File: Python/509. Fibonacci Number.py
class Solution:
    def fib(self, n: int) -> int:
        self.dp = [0]*(n+1)
        return self.solve(n)

    def solve(self,n):
        if n==0:
            return 0
        elif n==1:
            return 1
        if self.dp[n]!=0:
            return self.dp[n] 
        self.dp[n] = self.solve(n-1) + self.solve(n-2)
        return self.dp[n]